design/token-exchange-spi.md

```diff
@@ -0,0 +1,86 @@
+# Token Exchange SPI
+
+## Abstract
+This document outlines the design of the proposed Token Exchange SPI for Keycloak (TODO: JIRA link).
+## Rationale
+Keycloak does implement the OAuth 2.0 Token Exchange ([RFC 8693](https://tools.ietf.org/html/rfc8693)), but does that in a peculiar way (Securing Applications and Services Guide, [7. Token Exchange](https://www.keycloak.org/docs/latest/securing_apps/index.html#_token-exchange)):
+> Token exchange in Keycloak is a very loose implementation of the OAuth Token Exchange specification at the IETF. We have extended it a little, ignored some of it, and loosely interpreted other parts of the specification.
+
+Current implementation has the following limitations and shortcomings:
+* The following (optional) parameters are not supported:
+	* `resource`
+	* `scope`
+	* `actor_token` and `actor_token_type`
```

Comment:
These are related to the delegation scenarios covered by the specs that in our docs/implementation we refer to as impersonation.

From the list of use cases herein provided, is delegation among them?

Reply:
The biggest use case pushing us towards making this change is delegation, where the customer of a financial institution can delegate access to their accounts and data to a customer services representative from the FI. In this case, the customer is the subject of the requested token, and the CSR is the actor that will be using the token. In such scenarios, we have found that the customers and employees are sometimes using different OPs/IdPs and are therefore expected to use access tokens in different formats with different issuers. For an employee (who has a Ping token) to exchange that token for a customer (Keycloak) token without going to Ping to perform re-authentication, it should be possible to exchange the Ping token for a Keycloak token that has the appropriate claims embedded (e.g. the `act` claim identifying the employee)

Reply:
@jonm-bb This financial use-case you mentioned can be also solved using CIBA?

Reply:
CIBA would work from the target web application's perspective, which can contact Keycloak to trigger 'authentication', but as the employee will likely need to provide consent from one of the FI's web applications, it gets a bit awkward to contact for Keycloak to contact the web app for it to contact the user for them to 'authenticate' (by selecting which user they want to impersonate). For me, this use case is the wrong way round for CIBA - I would expect CIBA to be used when the person attempting to authenticate visits the app they are looking to authenticate into, which then contacts them over an appropriate channel. On top of that, the access token issued still needs to embed the `act` claim to identify the real end user. 
In this use case, the person being contacted (the CSR) is already authenticated elsewhere and needs to provide details of which customer they are being delegated access from; this is most seamless if they can simply exchange their access token for a limited one that will work with the target web app.
Other use cases where this token exchange mechanism would be advantageous is transparent token exchange at the system edge, where a token issued by the IdP is swapped for an internal format understood by the API being accessed, and CIBA is less useful there also.

Reply:
I see. I only mentioned CIBA because one of the use cases defined in the specs is related to an agent obtaining customer authorization to use his data for authentication/confirmation purposes. But yeah, not really delegation.

Thanks, for detailing the use case.